dev/tools/localization/gen_l10n_types.dart

```diff
@@ -0,0 +1,154 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+class L10nException implements Exception {
+  L10nException(this.message);
+
+  final String message;
+}
+
+class OptionalParameter {
+  const OptionalParameter(this.name, this.value) : assert(name != null), assert(value != null);
+
+  final String name;
+  final Object value;
+}
+
+class Placeholder {
+  Placeholder(this.resourceId, this.name, Map<String, dynamic> attributes)
+    : assert(resourceId != null),
+      assert(name != null),
+      example = _stringAttribute(resourceId, name, attributes, 'example'),
+      type = _stringAttribute(resourceId, name, attributes, 'type') ?? 'Object',
+      format = _stringAttribute(resourceId, name, attributes, 'format'),
+      optionalParameters = _optionalParameters(resourceId, name, attributes);
+
+  final String resourceId;
+  final String name;
+  final String example;
+  final String type;
+  final String format;
+  final List<OptionalParameter> optionalParameters;
+
+  bool get requiresFormatting => <String>['DateTime', 'double', 'int', 'num'].contains(type);
+  bool get isNumber => <String>['double', 'int', 'num'].contains(type);
+  bool get isDate => <String>['DateTime'].contains(type);
```

Comment:
nit: Could just be `type == 'DateTime'`, but I get why you might want to leave it like that. Just a suggestion.

Reply:
Done